File: specs/YU04-durable-control-feeds/generation/runtime-overrides/reference-data/src/stocks/source-epoch.repository.ts
import { Inject, Injectable, OnModuleInit } from '@nestjs/common';
import { Pool, RowDataPacket } from 'mysql2/promise';
import { DATABASE_POOL } from '../database/database.module';

/**
 * The security source epoch (ADR-021/ADR-019): a single row, seeded to 1 on first boot, bumped
 * only for a deliberate unrecoverable resync — never by normal operation.
 */
@Injectable()
export class SourceEpochRepository implements OnModuleInit {
  constructor(@Inject(DATABASE_POOL) private readonly pool: Pool) {}

  async onModuleInit(): Promise<void> {
    const [rows] = await this.pool.query<RowDataPacket[]>('select count(*) as cnt from stocks_source_epoch');
    if (Number(rows[0].cnt) === 0) {
      await this.pool.execute('insert into stocks_source_epoch (epoch) values (1)');
    }
  }

  async currentEpoch(): Promise<number> {
    const [rows] = await this.pool.query<RowDataPacket[]>('select epoch from stocks_source_epoch limit 1');
    return rows.length > 0 ? Number(rows[0].epoch) : 1;
  }
}

File: specs/YU04-durable-control-feeds/generation/runtime-overrides/account-service/src/main/java/finos/traderx/accountservice/outbox/SourceEpochRepository.java
package finos.traderx.accountservice.outbox;

import jakarta.annotation.PostConstruct;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

/**
 * The account source epoch (ADR-021/ADR-019): a single row, seeded to 1 on first boot, bumped
 * only for a deliberate unrecoverable resync — never by normal operation.
 */
@Repository
public class SourceEpochRepository {

  private final JdbcTemplate jdbcTemplate;

  public SourceEpochRepository(JdbcTemplate jdbcTemplate) {
    this.jdbcTemplate = jdbcTemplate;
  }

  @PostConstruct
  public void ensureSeeded() {
    Integer count = jdbcTemplate.queryForObject("select count(*) from account_source_epoch", Integer.class);
    if (count == null || count == 0) {
      jdbcTemplate.update("insert into account_source_epoch (epoch) values (1)");
    }
  }

  public long currentEpoch() {
    Long epoch = jdbcTemplate.queryForObject("select epoch from account_source_epoch limit 1", Long.class);
    return epoch == null ? 1L : epoch;
  }
}
